### PR TITLE
SD-439: handle orders on trader x application

### PR DESCRIPTION
**Pull request by domagojlatecki**
This PR attempts to handle orders which arrived via blockchain events in the case of trader application or InteractiveBrokers container restart. I tested this locally and it looks like it's working fine... May not be the most stable solution, but should be good enough for the demo.

### CODE REVIEW

File: src/main/kotlin/com/ampnet/tradeservice/repository/OrderRepository.kt [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+import org.springframework.stereotype.Repository
+import java.math.BigDecimal
+import java.time.Duration
+import java.time.OffsetDateTime
+
+@Repository
+@Suppress("TooManyFunctions")
+class OrderRepository(private val dslContext: DSLContext) {
+
+    fun storeBuyOrder(order: PlacedBuyOrder): SerialId<PlacedBuyOrder> {
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/tradeservice/repository/OrderRepository.kt:24]: You could do this function without return like:
```suggestion
    fun storeBuyOrder(order: PlacedBuyOrder): SerialId<PlacedBuyOrder> = dsl...
```

File: src/main/kotlin/com/ampnet/tradeservice/service/ib/wrappers/LoggingOrderWrapper.kt [hunk 2]
```diff
@@ -64,55 +84,15 @@ class LoggingOrderWrapper(private val blockchainService: BlockchainService) : EO
...
-            tokenAmount = Amount.fromSharesAmount(placedOrder.numShares).value,
-            wallet = placedOrder.wallet
-        )
-    }
-
     fun nextOrderId() = nextOrderId.getAndIncrement()
 
-    fun queueBuyOrder(order: PlacedBuyOrder) {
-        queuedOrders[order.interactiveBrokersOrderId.value] =
+    fun queueBuyOrder(order: SerialId<PlacedBuyOrder>) {
```
Comment (githinho) [inline src/main/kotlin/com/ampnet/tradeservice/service/ib/wrappers/LoggingOrderWrapper.kt:89]: I would rather put public functions before private functions.

File: src/main/kotlin/com/ampnet/tradeservice/queue/OrderQueueService.kt [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        BackgroundJob.scheduleRecurrently("fail-long-pending-orders", "*/60 * * * * *") {
+            failLongPendingOrders(orderProperties.maxPendingDuration)
+        }
+
+        BackgroundJob.scheduleRecurrently("resubmit-prepared-orders", "*/60 * * * * *") {
+            resubmitPreparedOrders(orderProperties.maxPreparedDuration)
+        }
+    }
+
+    fun failLongPendingOrders(minAge: Duration) {
```
[the remark this answers is not in the document]
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/tradeservice/queue/OrderQueueService.kt:34]: It cannot be private - that is the first thing I tried, but then JobRunr fails when it tries to schedule the job. So it needs to be public.

File: src/main/kotlin/com/ampnet/tradeservice/repository/OrderRepository.kt (hunk 1 above)
Comment (domagojlatecki) [reply src/main/kotlin/com/ampnet/tradeservice/repository/OrderRepository.kt:24]: Yeah I know... I'm kind of used to using `=` syntax only for very short functions which have no side-effects and `{}` for everything else. But I will use `=` here.

Comment (githinho) [reply src/main/kotlin/com/ampnet/tradeservice/repository/OrderRepository.kt:24]: No problem... It was just a suggestion, you can leave it for the next time.

**Review by githinho — APPROVED** [review]

**Pull request merged**